utils/SimpleJson.cpp

```diff
@@ -0,0 +1,410 @@
+{
+    if (!cursorIn || !cursorOut || cursorIn > cursorOut)
+        return SimpleJsonNone;
+    else
+        return SimpleJson(cursorIn, cursorOut - cursorIn + 1);
+}
+
+void SimpleJson::validateFast()
+{
+    if (!_cursorIn || !_cursorOut)
+    {
+        _type = Type::None;
+        return;
+    }
+    while (_cursorIn != _cursorOut && std::isspace(*_cursorIn))
+        _cursorIn++;
+    while (_cursorOut != _cursorIn && std::isspace(*_cursorOut))
+        _cursorOut--;
+    if (*_cursorIn == '{' && *_cursorOut == '}')
+    {
+        _type = Type::Object;
+    }
+    else if (*_cursorIn == '[' && *_cursorOut == ']')
+    {
+        _type = Type::Array;
+    }
+    else
+    {
+        _type = acquirePrimitiveType();
+    }
+}
+
+SimpleJson::Type SimpleJson::acquirePrimitiveType()
+{
+    auto cursor = findStringEnd(_cursorIn);
+    if (cursor == _cursorOut)
+    {
+        return Type::String;
+    }
+    cursor = findNullEnd(_cursorIn);
+    if (cursor == _cursorOut)
+    {
+        return Type::Null;
+    }
+    cursor = findBooleanEnd(_cursorIn);
+    if (cursor == _cursorOut)
+    {
+        return Type::Boolean;
+    }
+    cursor = findNumberEnd(_cursorIn);
+    char _buffer[33];
+    if (cursor == _cursorOut && size() < sizeof(_buffer))
+    {
+        strncpy(_buffer, _cursorIn, size());
+        int64_t intVal;
+        double floatVal, intPart;
+        auto readInt = sscanf(_buffer, "%" SCNd64, &intVal);
+        auto readFloat = sscanf(_buffer, "%lf", &floatVal);
+        if (readInt && readFloat)
+        {
+            // Return Float if we have fractional part. Otherwise Integer should suffice.
+            return modf(floatVal, &intPart) != 0 ? Type::Float : Type::Integer;
+        }
+        if (readInt && !readFloat)
+            return Type::Integer;
+        if (!readInt && readFloat)
+            return Type::Float;
+    }
+    return Type::None;
+}
+
+// Finds the end of the string.
+// start should point at opening " character.
+// Returns position of the closing " character.
+const char* SimpleJson::findStringEnd(const char* start) const
+{
+    const char* end = nullptr;
+    if (!start || '"' != *start)
+        return end;
+
+    auto cursor = start;
+    while (++cursor <= _cursorOut)
+    {
+        if ('\\' == *cursor)
+        {
+            cursor++;
+            continue;
+        }
+        if ('"' == *cursor)
+        {
+            end = cursor;
+            break;
+        }
+    }
+    return end;
+}
+
+// Finds the end of the object.
+// start should point at opening { character.
+// Returns position of the closing } character.
+template <char OPEN_CHAR, char CLOSE_CHAR>
+const char* SimpleJson::findEnd(const char* start) const
+{
+    const char* end = nullptr;
+    auto cursor = start;
+    if (!start || OPEN_CHAR != *start)
+        return end;
+
+    int level = 1;
+    while (++cursor < _cursorOut)
+    {
+        switch (*cursor)
+        {
+        case OPEN_CHAR:
```

Comment:
This is dangerous parsing of subobjects. If there is a string with a } in, we are toast.

Reply:
This is dangerous parsing of subobjects. If there is a string with a } in, we are toast.

Reply:
Method literally does finds end of the Object {}  or Array [] while "fast forwarding" all inside. We are not parsing anything, but we have to be careful about strings, thus if we detect '"' we skip until the end of the string taking into account escaping.